core/src/main/scala/kafka/server/ZookeeperLeaderElector.scala

```diff
@@ -121,13 +121,17 @@ class ZookeeperLeaderElector(controllerContext: ControllerContext,
      */
     @throws(classOf[Exception])
     def handleDataChange(dataPath: String, data: Object) {
+      var shouldResign = false
       inLock(controllerContext.controllerLock) {
         val amILeaderBeforeDataChange = amILeader
         leaderId = KafkaController.parseControllerId(data.toString)
         info("New leader is %d".format(leaderId))
         // The old leader needs to resign leadership if it is no longer the leader
-        if (amILeaderBeforeDataChange && !amILeader)
-          onResigningAsLeader()
+        shouldResign = amILeaderBeforeDataChange && !amILeader
+      }
+
+      if (shouldResign) {
```

Comment:
It seems the comment is missing? Can we add a short comment like "We need to resign as a controller out of the controller lock to avoid potential deadlock issue." This would help other developers understand with less effort why we do this.